metric_learn/base_metric.py

```diff
@@ -49,3 +39,23 @@ def transform(self, X=None):
       X = check_array(X, accept_sparse=True)
     L = self.transformer()
     return X.dot(L.T)
+
+
+class MahalanobisMixin(six.with_metaclass(ABCMeta)):
+  """Mahalanobis metric learning algorithms.
+
+  Algorithm that learns a Mahalanobis (pseudo) distance :math:`d_M(x, x')`,
+  defined between two column vectors :math:`x` and :math:`x'` by:
+  :math:`d_M(x, x') = \sqrt{(x-x')^T M (x-x')}`, where :math:`M` is the
+  learned square matrix.
+
+  Attributes
+  ----------
+  metric_: `np.ndarray`, shape=(n_features, n_features)
+      The learned Mahalanobis matrix.
+  """
+
+  @property
```

Comment:
I guess I saw the ``@property`` decorator as a way to enforce children to have a ``metric_`` attribute but I forgot that it would make ``metric_`` read-only. Indeed, I do not see any reason why it would be better to have it read-only, so maybe we could do it another way ? Like set a ``metric_`` argument to None in the ``__init__`` ? Or maybe after all there is no need to enforce it: we could just document it, for instance putting ``metric_`` in ``MahalanobisMixin``'s docstring under ``Attributes``, and doing it also for children Metric Learners, like for instance in scikit-learn ``BaseEnsemble`` documents an attribute ``estimators_`` that is also documented in ``RandomForestClassifier``(see  
https://github.com/scikit-learn/scikit-learn/blob/7a23b2493fbe094db2528114d75edc2e9739da17/sklearn/ensemble/base.py#L60-L85  and   https://github.com/scikit-learn/scikit-learn/blob/4d9a12d175a38f2bcb720389ad2213f71a3d7697/sklearn/ensemble/forest.py#L738-L862 )